test/forcetree_test.c

```diff
@@ -0,0 +1,346 @@
+        return +1;
+
+    return 0;
+}
+
+#define NODECACHE_SIZE 100
+
+/*This checks that the force tree in Nodes is valid:
+ * that it contains every particle and that each parent 
+ * node contains particles within the right subnode.*/
+static int check_tree(const int firstnode, const int nnodes, const int numpart)
+{
+    int tot_empty = 0, nrealnode = 0, sevens = 0;
+    for(int i=firstnode; i<nnodes+firstnode; i++)
+    {
+        struct NODE * pNode = &Nodes[i];
+        int empty = 0;
+        /*Just reserved free space with nothing in it*/
+        if(pNode->father < -1.5)
+            continue;
+
+        for(int j=0; j<8; j++) {
+            /*Check children*/
+            int child = pNode->u.suns[j];
+            if(child == -1) {
+                empty++;
+                continue;
+            }
+            assert_true(child < firstnode+nnodes);
+            assert_true(child >= 0);
+            /*If an internal node*/
+            if(child > firstnode) {
+                assert_true(fabs(Nodes[child].len/pNode->len - 0.5) < 1e-4);
+                for(int k=0; k<3; k++) {
+                    if(j & (1<<k))
+                        assert_true(Nodes[child].center[k] > pNode->center[k]);
+                    else
+                        assert_true(Nodes[child].center[k] <= pNode->center[k]);
+                }
+            }
+            /*Particle*/
+            else {
+                P[child].PI += 1;
+                /*Check in right quadrant*/
+                for(int k=0; k<3; k++) {
+                    if(j & (1<<k)) {
+                        assert_true(P[child].Pos[k] > pNode->center[k]);
+                    }
+                    else
+                        assert_true(P[child].Pos[k] <= pNode->center[k]);
+                }
+            }
+        }
+        /*All nodes should have at least one thing in them:
+         * maybe particles or other nodes.*/
+        if(empty > 6)
+            sevens++;
+        assert_true(empty <= 7);
+        tot_empty += empty;
+        nrealnode++;
+    }
+    assert_true(nnodes - nrealnode < omp_get_max_threads()*NODECACHE_SIZE);
+    for(int i=0; i<numpart; i++)
+    {
+        assert_true(P[i].PI == 1);
+    }
+    printf("Tree filling factor: %g on %d nodes (wasted: %d seven empty: %d)\n", tot_empty/(8.*nrealnode), nrealnode, nnodes - nrealnode, sevens);
+    return nrealnode;
+}
+
+static void do_tree_test(const int numpart)
+{
+    /*Sort by peano key so this is more realistic*/
+    #pragma omp parallel for
+    for(int i=0; i<numpart; i++) {
+        P[i].Key = PEANO(P[i].Pos);
+        P[i].Mass = 1;
+    }
+    qsort(P, numpart, sizeof(struct particle_data), order_by_type_and_key);
+    int maxnode = numpart;
+    assert_true(Nodes);
+    /*So we know which nodes we have initialised*/
+    for(int i=0; i< MaxNodes+1; i++)
+        Nodes_base[i].father = -2;
+    /*Time creating the nodes*/
+    struct timespec start, end;
+    clock_gettime(CLOCK_MONOTONIC,&start);
```

Comment:
MPI_Wtime is more portable than clock_gettime.

Reply:
Fixed in 3186c2b